pkg/proxyagent/agent/agent.go

```diff
@@ -194,6 +199,67 @@ func GetClusterProxyValueFunc(
 		if err != nil {
 			return nil, err
 		}
+
+		// Get agentIndentifiers and otherServices.
+		// agetnIdentifiers is used in `--agent-identifiers` flag in addon-agent-deployment.yaml.
+		// otherServices defines the services we want to expose to the hub.
+
+		// List all available managedClusterSets
+		managedClusterSetList := &clusterv1beta1.ManagedClusterSetList{}
+		err = runtimeClient.List(context.TODO(), managedClusterSetList)
+		if err != nil {
+			return nil, err
+		}
+
+		// deleted managedClusterSets are not included in the list
+		managedClusterSets := filterManagedClusterSetByDeletionTimestamp(managedClusterSetList.Items)
+
+		// only cluseterSet cover current cluster include in the list.
+		managedClusterSets, err = filterManagedClusterSetForCurrentCluster(managedClusterSets, cluster.Labels)
+		if err != nil {
+			return nil, err
+		}
+
+		managedClusterSetMap := map[string]clusterv1beta1.ManagedClusterSet{}
+		for _, managedClusterSet := range managedClusterSets {
+			managedClusterSetMap[managedClusterSet.Name] = managedClusterSet
+		}
+
+		// List all available serviceResolvers
+		serviceResolverList := &proxyv1alpha1.ManagedProxyServiceResolverList{}
+		err = runtimeClient.List(context.TODO(), serviceResolverList)
+		if err != nil {
+			return nil, err
+		}
+
+		// only legal serviceResolvers include in the list.
+		serviceResolvers := filterLegalManagedProxyServiceResolver(serviceResolverList.Items)
+
+		// deleted serviceResolvers are not included in the list
+		serviceResolvers = filterManagedProxyServiceResolverByDeletionTimestamp(serviceResolvers)
+
+		// filter serviceResolvers by managedClusterSet
+		serviceResolvers = filterManagedProxyServiceResolverWithManagedClusterSet(serviceResolvers, managedClusterSetMap)
+
+		// convert serviceResolvers to otherServices
+		otherServices := []map[string]string{}
+		for _, serviceResolver := range serviceResolvers {
+			otherServices = append(otherServices, convertManagedProxyServiceResolverToOtherService(cluster.Name, serviceResolver))
+		}
+		// remove dup and sort otherServices
+		otherServices = removeDupAndSortOtherServices(otherServices)
+
+		// add default agentIdentifiers
+		var aids []string
+		aids = append(aids, fmt.Sprintf("host=%s", cluster.Name))
+		aids = append(aids, fmt.Sprintf("host=%s.%s", cluster.Name, config.AddonInstallNamespace))
+
+		// add otherServices into aids
+		for _, otherService := range otherServices {
+			aids = append(aids, fmt.Sprintf("host=%s", otherService["url"]))
```

Comment:
if we only care about url here, why we need otherServices as a []map[string]string?
can we make it sets.String, so you will not have duplicates easily

Reply:
Both `urls` and `otherServices(now servicesToExpose)` is used to render manifest template.
URL is used in args `--agentIndentifiers` in agent deployments.
`servicesToExpose` is used in `pkg/proxyagent/agent/manifests/charts/addon-agent/templates/other-services.yaml` to generate `ExternalName` type services.

Reply:
can we make it sets.String, so you will not have duplicates easily.

I'm not sure because `servicesToExpose` is finally used to render a manifest template, and an `array` makes this part easy.